### PR TITLE
Add Typescript

### PR DESCRIPTION
**Pull request by ThieryMichel**
Add type for external api.

### CODE REVIEW

File: src/Validation.ts [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+        fn: (
+            invalidResult: InvalidResult,
+            index: number,
+            all: InvalidResult[]
+        ) => InvalidResult
+    ) => Valid;
+    getResult: () => void;
+}
+
+export interface Invalid {
```
Comment (fzaninotto) [inline src/Validation.ts]: This type name is a bit mysterious. Invalid what?

File: src/Validation.ts [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+const isInvalid = (
+    value: Valid | Invalid | AsyncValidation
+): value is Invalid => value.isValid === false;
+
+const isAsync = (
+    value: Valid | Invalid | AsyncValidation
+): value is AsyncValidation => !!value.fork;
+
+// Valid will hold no value
+export const Valid = (): Valid => ({
```
Comment (fzaninotto) [inline src/Validation.ts]: having both a function and a type with the same name is confusing IMHO

File: src/Validator.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
+import {
+    asyncLift,
+    lift,
+    AsyncValidation,
+    SyncValidation,
+    InvalidResult,
+} from "./Validation";
+
+type run<T extends SyncValidation | AsyncValidation> = (
```
Comment (fzaninotto) [inline src/Validator.ts]: First time I see a type that doesn't start with a capital letter

File: src/utils.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
+import { Validator } from "./Validator";
+import { SyncValidation, AsyncValidation } from "./Validation";
+
+export const and = (validator1: any, validator2: any) =>
+    validator1.and(validator2);
+
+export const path = (keys: string[], obj: Object) =>
```
Comment (fzaninotto) [inline src/utils.ts:14]: a JSDoc with sample input/output would help understand what this function does

File: src/validators/object.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+        return extraneousKeys.length > 0
+            ? `Value has extraneous keys: ${extraneousKeys.join(", ")}`
+            : undefined;
+    });
+
+export const isExactObject = (keys: string[]) =>
+    isObject.and(hasNoExtraneousKeys(keys));
+
+interface shape {
```
Comment (fzaninotto) [inline src/validators/object.ts]: same remark, why not use a capital S?

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: src/Validation.ts (hunk 2 above)
Comment (ThieryMichel) [reply src/Validation.ts]: Since the function is a factory to generate the type, I do not see an issue here.

File: src/Validation.ts (hunk 1 above)
Comment (ThieryMichel) [reply src/Validation.ts]: It's an Invalid Validation. It represents an Invalid validation result.
It is also part of the monoid Validation.
A monoid is a set of value that can be combined together by an operation into other member of the same set.
Here the set of Validation is comprised of Valid, Invalid and Async.
IMO Invalid is the right name.

**Pull request merged**